Lib/unittest/case.py

```diff
@@ -1483,3 +1486,149 @@ def shortDescription(self):
+    is stopped and closed.
+
+    When subclassing AsyncioTestCase, the event loop is available
+    via the ``loop`` property.  Tests can assume that `self.loop`
+    refers to the event loop created for the test case.  It is also
+    the active event loop as returned by ``asyncio.get_event_loop()``.
+    Test cases SHOULD NOT change the active event loop during the test
+    run.
+
+    The lifecycle of a single test execution is thus::
+
+        - ``TestCase`` instance is created
+        - ``setUpClass`` is executed
+        - for each test method
+          - new event loop is created
+          - ``setUp`` is executed on the event loop
+          - if ``setUp`` succeeds
+            - the test method is run on the event loop if it is a
+              co-routine; otherwise it is simply called
+            - ``tearDown`` is executed on the event loop
+          - the event loop is stopped if it is somehow running
+          - ``doCleanups`` is called
+            - callables registered with ``addCleanup`` are called
+              in reverse-order executing them on the event loop if
+              necessary
+            - the event loop is stopped if necessary
+            - the event loop is closed and unregistered
+
+    The process is tightly managed by the ``_runTest`` and ``doCleanups``
+    methods.  Care must be taken if extending this class to ensure that
+    the event loop is properly managed.
+
+    """
+
+    def __init__(self, methodName='runTest'):
+        super().__init__(methodName)
+        self.__loop = None
+
+    @classmethod
+    def setUpClass(cls):
+        """Hook method for one-time initialization.
+
+        Subclasses MUST invoke this method when extending it.
+        """
+        super().setUpClass()
+        cls.__saved_policy = asyncio.events._event_loop_policy
+
+    @classmethod
+    def tearDownClass(cls):
+        """Hook method for one-time cleanup.
+
+        Subclasses MUST invoke this method when extending it.
+        """
+        super().tearDownClass()
+        asyncio.set_event_loop_policy(cls.__saved_policy)
+
+    async def asyncSetUp(self):
+        """Hook method for setting up the test fixture before exercising it.
+
+        This method invokes ``setUp`` inline so subclasses MUST invoke this
+        method using the ``await`` keyword when extending it.
+        """
+        self.setUp()
+
+    async def asyncTearDown(self):
+        """Hook method for deconstructing the test fixture after testing it.
+
+        This method invokes ``tearDown`` inline so subclasses MUST invoke
+        this method using the ``await`` keyword when extending it.
+        """
+        self.tearDown()
+
+    @property
+    def loop(self):
+        """Active event loop for the test case."""
+        if self.__loop is None:
+            self.__loop = asyncio.new_event_loop()
+            self.__loop.set_debug(True)
+            asyncio.set_event_loop(self.__loop)
```

Comment:
Does the test case have to modify the global state like this? It is calling `self.loop.run_until_complete` when running things.

Reply:
I think that it does.  This is the only way to ensure that the code under test gets the correct loop.